feat: support lazy loading services

pimlie opened the pull request:
Resolves: #231 

This PR includes a breaking change, but that could be mitigated by making legacyMode + injectModule default true. Will add docs after first review.

This add 3 new options:
- `Bool legacyMode`: to inject each service separately, default will now be a single `$fire` prop
- `Bool injectModule`: whether to inject the modules/objects or not (works also in legacyMode)
- `Bool lazy`: whether to use lazy mode

Its _not_ supported to lazy load specific services. Its either all or nothing.

I didnt touch the other plugins, could some help with testing those. Or maybe we can merge them into the service template?

lupas commented:
Hey @pimlie 

This looks awesome, thanks a lot! Unfortunately won't have time to have a closer look the next 2-3 days - I'll try to review and test it either on Wednesday or on Thursday.

You'll hear from me!

lupas commented on lib/plugins/main.js:
```diff
@@ -1,361 +1,141 @@
-
+    <% for (const service of clientServices) { %>
+    inject('fire<%= options.injectionMapping[service] %>', <%= getServiceName(service) %>)
+    inject('fire<%= options.injectionMapping[service] %>Obj', firebase.<%= service %>)
```
```suggestion
    inject('fire<%= options.injectionMapping[service] %>Obj', firebase.<%= options.serviceMapping[service] %>)
```

Same as above. Does't make a difference here because the keys and the values of the clientServices are exactly the same in the serviceMapping, but that's just luck.

lupas commented:
> This PR includes a breaking change, but that could be mitigated by making legacyMode + **injectModule** default true.

Why do you think "injectModule" default `true` will mitigate breaking changes? As far as I can see we never injected `$fireObj` before. What am I missing?

pimlie commented on lib/template-utils.js:
```diff
@@ -0,0 +1,98 @@
+  storage: 'Storage',
+  functions: 'Func',
+  messaging: 'Mess',
+  performance: 'Performance',
```
Thx, this wasnt an intended change actually :)

pimlie commented:
Thanks for review, have addressed some of your comments. Your proposal for v7/v8 sound good to me, have added a check + info log for this but feel free to adjust the message to your liking.

- When would the module fail to initialize? The idea of the legacyMode is to prevent breaking changes, so not sure what you suggest we'd add (the basic stuff we should add proper tests for)
- have changed the path in movePluginBeforeAuthHelper but I would suggest to deprecate this method. We could use the [`builder:extendPlugins`](https://github.com/nuxt/nuxt.js/blob/dev/packages/builder/src/builder.js#L275) to provide similar functionality
- `plugins/initAuth` why is this a separate plugin? cant we move this into the service template?
- The injectModule check was missing for the legacyMode, have added that for legacyMode too

Some other things I noticed:
- We could actually remove all those xxReady vars and just check fire.xx immediately
- Not sure if those writeImport helpers are really beneficial in the service templates
- There might still be an issue with remoteConfig due to the dash in the name

Feel free btw to pair-up with me on this pr if you want and have time, you might be quicker with checking/fixing some stuff then I do as I only have used auth & realtimeDb up to now ;)

lupas commented on lib/module.js:
```diff
@@ -11,13 +11,14 @@ module.exports = function (moduleOptions) {
   const currentEnv = getCurrentEnv(options)
 
+  // TODO: Remove with v8
+  if (this.nuxt.options.dev && !(this.options.firebase && 'legacyMode' in this.options.firebase) && !('legacyMode' in moduleOptions)) {
```
didn't really mean to move it down, it's rather about the `this.options.firebase` that can be undefined

lupas commented:
> Thanks for review, have addressed some of your comments. Your proposal for v7/v8 sound good to me, have added a check + info log for this but feel free to adjust the message to your liking.

Great, perfectly fine like that!

> * When would the module fail to initialize? The idea of the legacyMode is to prevent breaking changes, so not sure what you suggest we'd add (the basic stuff we should add proper tests for)

What I meant is: When legacyMode is false by default, users who updated to the new version would instantly face a random error message because they haven't turned legacyMode on yet. But since it's true by default now this has been solved.

> * have changed the path in movePluginBeforeAuthHelper but I would suggest to deprecate this method. We could use the [`builder:extendPlugins`](https://github.com/nuxt/nuxt.js/blob/dev/packages/builder/src/builder.js#L275) to provide similar functionality

Do you mean by letting a user provide an array of plugin-names (in regex) via config which the user wants to be placed after the firebase/index plugin but before the initAuth plugin, and we would then slide these plugins in between? Would that work?

> * `plugins/initAuth` why is this a separate plugin? cant we move this into the service template?

Because we have situations where we want a plugin to be called after firebase init but before initAuth, see for example https://github.com/nuxt-community/firebase-module/issues/146.

Another example, I often use `this.$sentry.configureScope` within `onAuthStateChanged`, which would be undefined if the sentry plugin is loaded after the initAuth plugin. So with the helperFunction I can move it in between easily.

> * We could actually remove all those xxReady vars and just check fire.xx immediately

Honestly, I never wrote a line of that ready-code so I'm not entirely sure what it is there for :P Totally up for cleaning that up a bit 👍 

> * Not sure if those writeImport helpers are really beneficial in the service templates

Which one you mean? Would you think it's easier if we hardcode the imports? Problem is that with all the import options  we have (static/chunkName/WebpackComments) this would be rather verbose if we don't use the helper in each service template, no?

> * There might still be an issue with remoteConfig due to the dash in the name

It has to be imported like `import 'firebase/remote-config'`, where do you see the issue?

> Feel free btw to pair-up with me on this pr if you want and have time, you might be quicker with checking/fixing some stuff then I do as I only have used auth & realtimeDb up to now ;)

I'll try to do as much as I can, but have people over for the next 10 days so won't have much time to code, after next week I hopefully can do more :)

pimlie commented:
> RE: extendPlugins

Yes, exactly like that! That hook returns the full plugins array after resolving, we just have to make sure to re-order in-place. Ie we cannot return a new array but have to use the one the hook receives. Another option would be to copy what nuxt/auth does: https://auth.nuxtjs.org/recipes/extend.html

> RE: initAuth

Aha, thx for explaining. So this plugin actually doesnt work atm for lazy mode, because at plugin exec firebase is likely not loaded yet (unless someone adds a plugin before that awaits ready it). If initAuth needs to be run as last, we could just use the extendPlugins hook to always put it as the last plugin. Within the firebase plugin we could then provide an awaitable injection that users can call, they would just need to make sure to never await that injection within their plugin because that would never resolved (ie not `await authChanged()` but `authChanged().then(() => `

> RE: xxReady

Yeah sorry, I was commenting on my own code. Progressive insights ;)

> RE: template-utils

Yeah, but that extra verbosity could improve code readability, eg `if (options.lazy) import` is easier to understand then following a function call that has some inner magic. All those templates just arent ideal, but we cant do much about them for nuxt v2 :(

lupas commented:
@pimlie 
## RE: extendPlugins

I'm all in for doing it the way other modules do it, too.

**OR** we could also just remove it entirely and write a short note in the docs on what needs to be done to move a plugin in between -> that way the user has maximal flexibility and can do it the way he/she wants to.

For example, in my projects I do this:
```js
extendPlugins(plugins) {
      const reorderArray = [
        'firebase-module/main',
        'setupServices', // Private Plugin
        'sentry.server', // Needs to be before initAuth
        'sentry.client', // Needs to be before initAuth
        'initAuth',
      ]
      nuxtPluginMover(plugins, reorderArray)
      return plugins
    },
```
Where nuxtPluginMover is a helper function (not related to nuxt/firebase) that just reorders plugins the way you input them, but doesn't expect you to input all plugins - just the one's you feed it will be re-ordered.

Could publish that nuxtPluginMover if it is helpful.

That way, in case the user needs to reorder plugins due to some other reason, we don't mess up stuff.

## RE: initAuth
Wouldn't all plugins be initialized on page load anyway, so if we lazy load auth e.g. only after a route change the plugin initialization would fail earlier anyway, even though initAuth is set to be the last plugin? I might be misunderstanding something here.

Or is there a way to lazyLoad a plugin in Nuxt? So if the user uses the lazy loaded auth-module, he/she could load the plugin after that?

```js
this.$router.push('/authPage')
// then in the authPage component we load auth and could after that load the initAuth plugin?
this.$nuxt.loadPlugin('initAuth.js') // something like that?
```

## RE: template-utils
You might be right. Making it more readable is imho worth more than having a magic function that strips away every little repetitive string. Go for it!

In general, if you see something that you feel can be cleaned up, feel free to do it. The code grew a lot in the last years without me finding much time to clean it up, and you have some more JS experience than me so I fully trust you here ;)

pimlie commented:
Have fixed initAuth, decided to put the subscribe code in the auth server template to prevent having to switch using $fire.auth or $fireAuth. Also I changed the interface a  bit (but legacy mode should work the same), you can now subscribe & unsubscribe as many times as you want.

I dont have a strong opinion about all the template stuff because I dont have a need for it atm, if I would have to choose then I would do the same as nuxt/auth v5 does (not sure that that is). But I think we can leave that out of the scope of this pr.

Also other rewrites (ie template-utils) are out of scope for this pr. I mean, it should already work so why change it again. For nuxt v3 we would probably need to refactor anyway.

lupas commented:
@pimlie 
Alright, let's not bother about the rewrites at this point and try to get this release-ready asap instead.

I just tried your changes with nuxt-fire-demo in legacyMode and am now getting the following error with my unchanged legacy config:
```
client.js?06a0:96 Error: inject('fireConfigObj', value) has no value provided
    at _inject (index.js?f26e:145)
    at _callee$ (index.js?d90b:77)
    at tryCatch (runtime.js?96cf:63)
    at Generator.invoke [as _invoke] (runtime.js?96cf:293)
    at Generator.eval [as next] (runtime.js?96cf:118)
    at asyncGeneratorStep (asyncToGenerator.js?1da1:3)
    at _next (asyncToGenerator.js?1da1:25)
```

and

```
client.js?06a0:96 Error: inject('fireAnalytics', value) has no value provided
    at _inject (index.js?f26e:143)
    at _callee$ (index.js?d90b:27)
    at tryCatch (runtime.js?96cf:63)
    at Generator.invoke [as _invoke] (runtime.js?96cf:293)
    at Generator.eval [as next] (runtime.js?96cf:118)
    at asyncGeneratorStep (asyncToGenerator.js?1da1:3)
    at _next (asyncToGenerator.js?1da1:25)
```

But latter for some weird reason **only**, if I delete all other services and only have this one enabled. Feels like some sort of race-condition, so that if x other services get injects first suddenly firebase.analytics is not undefined anymore, but if it's at the top of the code it throws this issue. So it seems it doesn't wait for firebase.analytics() to complete...

Tried to figure out the issue but couldn't find it in the short time I had and gotta run now, hope I'll find some time the coming week.

Can you see where this issue might come from?

topmusica00 commented:
client.js?06a0:96 Error: inject('fireConfigObj', value) has no value provided
    at _inject (index.js?f26e:145)
    at _callee$ (index.js?d90b:77)
    at tryCatch (runtime.js?96cf:63)
    at Generator.invoke [as _invoke] (runtime.js?96cf:293)
    at Generator.eval [as next] (runtime.js?96cf:118)
    at asyncGeneratorStep (asyncToGenerator.js?1da1:3)
    at _next (asyncToGenerator.js?1da1:25)
Roxa dão Free Fire Gelo Agachando
#FreeFire

lupas commented:
Unfortunately super busy until September, I'll try to have a closer look at the PR then. Would love to get this merged before there's too many changes on the master branch.

br

pimlie commented:
@lupas Yeah, same here unfortunately :( Sorry about that, will get back to this when I can.

lupas commented:
Finally found some time to properly look through the code & work on the PR - lazy mode is awesome 🥳
Did some changes today and will continue tomorrow, but need sleep now 🥱

- Merged in the current dev version (nuxt-content docs 🥳). Ready to add adjust the docs.
- Fixed quite a bunch of issues that I found
- Various cleanups
- Simplified a bunch of the naming stuff - don't think we need to dynamically generate the names if they are static anyway. The way it is now it is imho much more comprehensible and easier to manage.
- Use exactly the same names for injection as firebase uses. So firebase.remoteConfig -> $fire.remoteConfig. Less confusing than "$fire.config". 

I quickly smoke-tested basic functionality:
- legacyMode, all services ✅
- lazy=false, all services ✅
- lazy=true, all services ✅

Ran into issues with initAuth with lazy mode, will have a look at it tomorrow.

Next steps:
- [ ] Adjust tests for most recent changes
- [ ] Test everything properly & adjust
- [ ] Documentation

lupas commented:
# Open Issues
- **Issue 1:** initAuth with lazy mode still does not work (or I broke it again...)
   ```js
   // nuxtServerInit
   this.$fire.auth cannot be accessed
   ```

# Open ToDo's

- **ToDo 1:** Adjust tests for most recent changes (@lupas)
- **ToDo 2:** Test every specific use case properly manually (@lupas)
- **ToDo 3:** We could actually remove all those xxReady vars and just check fire.xx immediately (@pimlie)
- **ToDo 4:** Adjust Typescript `index.d.ts` (add `$fire.fooReady()`) (@lupas) (depends on ToDo 3)
- **ToDo 6:** Deprecate movePluginBeforeAuthHelper and use builder:extendPlugins instead with a config e.g. `movePluginsBeforeAuthHelper: ['regex123','regex456']`
- **ToDo 5:** Documentation (@pimlie & @lupas)

# Questions
- **Question 1:** @pimlie
Do we even need injectModule for legacyMode? If we remove legacy mode with v8 anyway, I think legacyMode should just reflect current v6. Adding new functionality just complicates things. Like this, users using legacyMode will not additionally have to add injectModule:true
- **Question 2:** @pimlie
I thought about it again and think it would be best to have the following defaults:
   ```js
   // v7
   legacyMode: false,
   injectModule: false  // no effect on legacy mode, there we always inject it (as used to in v6)
   // v8
   // legacyMode gone
   injectModule: false
   ```
   When `legacyMode` config is not set, we inform about the breaking changes and that people should either migrate and set flag to `false` or use legacyMode which will be no breaking changes to v6.
- **Question 3:** @pimlie
What do you think about having `injectModule` as a config on service level? Currently, we either inject it for all services or none. But in my experience it's frequently used oly for certain services like `auth`, whereas in other services it's rarely used.

lupas commented:
Made a bold move and **removed all legacyMode functionality**.

Imho the breaking changes are not that difficult to adapt to, and I don't see much benefit in having a complicated transition period where some users will be on legacyMode and others not. I think one has quickly renamed $fireFoo to $fire.foo.

I'll next start writing the documentation and do further tests (ssr / serverLogin etc.).

@pimlie do you think you could quickly checkout the Question 1 above and let me know if it would make sense to break this down per-service? I don't want to include all modules if not necessary - except that makes no difference performance-wise.

pimlie commented:
Hey, thanks for all your work on this! I have very little time available atm to work on this unfortunately, but will try to help when I can.

+1 for removing legacy mode, it would have been nice tho to allow for some transition period but a good migration guide should be ok too. Agree the changes arent that difficult to implement (altho it could be many places where you'd have to change this in your codebase). But they can always pin their version I guess

Im not rly sure if there will a performance benefit or not with removing `injectModule`, it depends how the Firebase code itself is structured (ie if it has multiple entry points) and if the GC can release certain parts of the lib when you dont use them. I will leave this decision up to you, but personally I think it makes sense to support a per service setting (ie only include when you rly need it)

If the stringify function issue happens within a template then it can be resolved by using `serialize` or `serializeFunction` that is available from Nuxt's template scope (ie using `<%= serializeFunction(option.var) =>`)

lupas commented:
@pimlie 
Thanks for the quick response!

No worries at all, same here, hard to find enough time to finish this up but I think I'll get it done soon :)

I'll make sure to write a proper migration guide and then I think this will be fine without legacy mode.
About the injectModule I will make some test and see how it performs - I guess it'll make sense to split it up.

You wrote further up "_We could actually remove all those xxReady vars and just check fire.xx immediately_".
What exactly did you mean by that? Do you see a way to *automatically* load the services the first time they are used without us having to call `ready()`? (that would be awesome)

lupas commented:
**Remaining tasks:**

- [ ] **injectModule**: Test if it makes difference regarding performance, either inject always everything, or per service individually. If it doesn't make a difference at all, maybe even delete the option to set it true/false?
- [ ] **stringify issue**: Check if it comes from changes in module or if its just me doing something wrong in in the demo, if it's here -> fix it as mentioned above
- [ ] **tests**: Add additional tests for all scenarios and update demo to test every scenario manually

Other than that documentation already looks good to me and it seems to work quite nicely.

lupas commented:
@pimlie 
I knew what lead to my demo showing "Cannot stringify a function":
It was having this empty fetch() function defined in a vue component:

```js
  fetch() {
    // INFO -> this.$fire.auth user etc. are accessible in fetch
    // INFO -> this.$store.state.authUser is accessible even on server-side in fetch
  },
```
<img width="660" alt="Screenshot 2020-10-24 at 21 15 05" src="https://user-images.githubusercontent.com/4020037/97091694-fdb13f80-163d-11eb-80d3-374dceeffd15.png">

Removing that got rid of the warning.

Just FYI, I think it's quite interesting. Couldn't reproduce it on a new nuxt repo, so not sure what exactly is the issue here..

lupas commented:
@pimlie 
Finished the PR and merged into dev, tested everything and it all seemed to work perfectly.

I'll do some more improvements to the module and merge another PR, but then will release all this as v7.0.0 in the coming days.

Thanks a lot for your support and this PR, this is great! =)

pimlie commented:
@lupas Thank you so much for finishing this pr! :100:
